### PR TITLE
Spark: Fix IllegarlArgumentException when filtering on BinaryType column

### PR DESCRIPTION
**Pull request by izchen**
The test case reproducing the exception：

```sql
CREATE TABLE filter_binary_test (cbinary BINARY) USING ICEBERG;
SELECT cbinary FROM filter_binary_test WHERE cbinary > X'110F';
```

```
java.lang.IllegalArgumentException: Cannot convert bytes to SQL literal: java.nio.HeapByteBuffer[pos=0 lim=2 cap=2]
```

Like this SQL, BinaryLiteral(ByteBuffer) is not an illegal value.

in addition, the BinaryLiteral string format in this PR is the same as spark.

https://github.com/apache/spark/blob/b874bf5dca4f1b7272f458350eb153e7b272f8c8/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala#L347

### CODE REVIEW

**Comment by izchen** [comment]
Related issue: https://github.com/apache/iceberg/issues/2934

**Comment by izchen** [comment]
@rdblue @kbendick , could you help to review this PR? :)

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java [hunk 1]
```diff
@@ -639,5 +647,12 @@ Object writeReplace() throws ObjectStreamException {
     protected Type.TypeID typeId() {
       return Type.TypeID.BINARY;
     }
+
+    @Override
+    public String toString() {
+      byte[] binary = new byte[value().remaining()];
+      value().duplicate().get(binary);
```
Comment (kbendick) [inline api/src/main/java/org/apache/iceberg/expressions/Literals.java:654]: Same note as above.

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java [hunk 2]
```diff
@@ -639,5 +647,12 @@ Object writeReplace() throws ObjectStreamException {
...
+      return "0x" + BaseEncoding.base16().encode(binary);
```
Comment (kbendick) [inline api/src/main/java/org/apache/iceberg/expressions/Literals.java:655]: I'm not sure if we have any other places where we convert a `ByteBufer` to a string.

If there are any, it seems like potentially we should add the stringify function to the `ByteBuffer` utility class and then try to use it uniformly everywhere. You don't necessarily need to update the other places to use them in this PR, but would you mind taking a look and seeing if you find any?

Comment (kbendick) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:655]: Follow up question: Are there cases where people might not use base16 for the representation or where the leading zero might not be used? I don't think so for Spark, but I'm less sure with Flink for example.

For Spark, looking at the documentation, it seems like it's just `X`, without the leading zero: https://spark.apache.org/docs/latest/sql-ref-literals.html#binary-literal

I'm not as sure about Trino, Flink, and other systems though.

For Trino, it looks to also start with the capital `X` (without the zero): https://trino.io/docs/current/language/types.html#varbinary

Looking it up, for Flink, it looks like maybe it's actually represented as `BINARY(n)` or `VARBINARY(n)`. This is the one I'm least sure about: https://nightlies.apache.org/flink/flink-docs-release-1.13/docs/dev/table/types/#binary

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java [hunk 3]
```diff
@@ -636,17 +635,17 @@ public String or(String leftResult, String rightResult) {
...
+          return pred.ref().name() + " > " + pred.literal();
         case GT_EQ:
-          return pred.ref().name() + " >= " + sqlString(pred.literal());
+          return pred.ref().name() + " >= " + pred.literal();
         case EQ:
-          return pred.ref().name() + " = " + sqlString(pred.literal());
+          return pred.ref().name() + " = " + pred.literal();
         case NOT_EQ:
-          return pred.ref().name() + " != " + sqlString(pred.literal());
+          return pred.ref().name() + " != " + pred.literal();
```
Comment (kbendick) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:648]: Why did we get rid of the `sqlString` function here? Is there a change that removes the need to quote strings for example?

Also, since the representation of `ByteBuffer` is potentially engine specific, would it be better to add the conversion to the `sqlString` here for Spark (e.g. using the leading capital X, like `X'123456'`).

This would allow other engines to handle it themselves (e.g. if Flink doesn't use leading `X` format).

I worry about unintended side effects from removing the explicit quoting as I've seen more PRs come up that handle adding and removing quotes as needed, but possibly you're more familiar with those conversations than I am.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java [hunk 4]
```diff
@@ -39,6 +39,7 @@
 import org.junit.AfterClass;
 import org.junit.Assert;
 import org.junit.BeforeClass;
+import org.junit.internal.ExactComparisonCriteria;
```
Comment (kbendick) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java:42]: Is it normal to use this class that's got `internal` in the package name? Do we have to worry about behavioral changes if we upgrade JUnit?

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java [hunk 5]
```diff
@@ -599,6 +600,13 @@ public String toString() {
     Object writeReplace() throws ObjectStreamException {
       return new SerializationProxies.FixedLiteralProxy(value());
     }
+
+    @Override
+    public String toString() {
+      byte[] binary = new byte[value().remaining()];
+      value().duplicate().get(binary);
```
Comment (kbendick) [inline api/src/main/java/org/apache/iceberg/expressions/Literals.java:607]: There's an existing ByteBuffers utility class, org.apache.iceberg.util.ByteBuffers, which you can use here. Can you please use that for consistency? It's also more efficient in some cases (doesn't always necessarily allocate, handles offsets more thoroughly, etc).

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java [hunk 6]
```diff
@@ -659,17 +658,7 @@ public String or(String leftResult, String rightResult) {
...
-
-    private static String sqlString(org.apache.iceberg.expressions.Literal<?> lit) {
-      if (lit.value() instanceof String) {
-        return "'" + lit.value() + "'";
-      } else if (lit.value() instanceof ByteBuffer) {
-        throw new IllegalArgumentException("Cannot convert bytes to SQL literal: " + lit);
-      } else {
-        return lit.value().toString();
-      }
+      return literals.stream().map(Object::toString).collect(Collectors.joining(", "));
```
Comment (kbendick) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:661]: This seems possibly incorrect or a deviation from the behavior before.

It looks like this code is calling Object.toString on `Literal<T>`, whereas before, the `toString` call came from `Literal<T>::value`. Is that the intended change?

**Review by kbendick — COMMENTED** [review]
Thanks for your contribution @izchen!

Left some feedback. I have some concerns around the removal of `sqlString` function for literals as I don't see any corresponding changes for strings.

Eventually, we'll need to update more than just Spark v3.2. Additionally, it would be nice if Flink and even MR could be handled in this PR. But I'm especially interested in Flink, as the documentation seems that it uses the constructor looking syntax. I admit I'm not familiar enough with using byte literals in Flink SQL to be sure though.

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 2 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java:655]: This should just be for debugging. If we are converting a literal to a string and back, then that's a problem.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java (hunk 3 above)
Comment (rdblue) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:648]: @kbendick is right. This should not remove the `sqlString` function. That converts a literal into a form that is usable by Spark. Instead, you should update that function to produce Spark's binary literal format.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java [hunk 7]
```diff
@@ -157,7 +158,11 @@ private void assertEquals(String context, Object[] expectedRow, Object[] actualR
       Object actualValue = actualRow[col];
       if (expectedValue != null && expectedValue.getClass().isArray()) {
         String newContext = String.format("%s (nested col %d)", context, col + 1);
-        assertEquals(newContext, (Object[]) expectedValue, (Object[]) actualValue);
+        if (expectedValue.getClass().getComponentType().isPrimitive()) {
+          new ExactComparisonCriteria().arrayEquals(newContext, expectedValue, actualValue);
+        } else {
+          assertEquals(newContext, (Object[]) expectedValue, (Object[]) actualValue);
+        }
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java:165]: Why was this needed?

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/sql/TestSelect.java [hunk 8]
```diff
@@ -49,8 +49,8 @@ public TestSelect(String catalogName, String implementation, Map<String, String>
 
   @Before
   public void createTables() {
-    sql("CREATE TABLE %s (id bigint, data string, float float) USING iceberg", tableName);
-    sql("INSERT INTO %s VALUES (1, 'a', 1.0), (2, 'b', 2.0), (3, 'c', float('NaN'))", tableName);
+    sql("CREATE TABLE %s (id bigint, data string, float float, binary binary) USING iceberg", tableName);
+    sql("INSERT INTO %s VALUES (1, 'a', 1.0, X''), (2, 'b', 2.0, X'11'), (3, 'c', float('NaN'), X'1111')", tableName);
```
Comment (rdblue) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/sql/TestSelect.java:53]: Rather than repurposing old test cases, can you create new ones? We want to avoid mixing tests together in confusing ways that look like other failures.

Comment (kbendick) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/sql/TestSelect.java:53]: I'd also add that repurposing existing test cases makes it harder for people who maintain forks and might not cherry-pick every commit.

File: api/src/main/java/org/apache/iceberg/util/ByteBuffers.java [hunk 9]
```diff
@@ -58,6 +59,11 @@ public static ByteBuffer copy(ByteBuffer buffer) {
     return ByteBuffer.wrap(copyArray);
   }
 
+  public static String encodeHexString(ByteBuffer buffer) {
+    byte[] bytes = toByteArray(buffer.duplicate());
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/util/ByteBuffers.java:63]: No need for `duplicate` here because `toByteArray` won't modify the buffer.

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 2 above)
Comment (izchen) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java]: This should just be for debugging.

In the base class, the toString method is `String.valueOf(value)` (In StringLiteral, it is `"\"" + value() + "\""`).

For other types of literals, the return value of toString is clear. But for ByteBuffer, the return value is `java.nio.HeapByteBuffer[pos=0 lim=16 cap=16]`.

So I implemented the literal toString method of  ByteBuffer type, for logging.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java (hunk 3 above)
Comment (izchen) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:648]: `sqlString` is implemented and used in the DescribeExpressionVisitor class.

I think this class is just for generating description information, and the description will be displayed in the log and Spark UI for debugging. So in my opinion, it is an acceptable way to use Literal.toString directly, which ensures that the description is the same as the style of other logs.

In fact, except for the difference between StringLiteral's single and double quotes, the current sqlString and Literal.toString are the same.

But if you think that keeping sqlString is a better way, I can also modify the code.
In addition, there are some differences between the description form of spark here, for example, the StringLiteral of spark does not have quotation marks. If you think that this should be the same as the behavior of spark, I can also modify the code.

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java (hunk 4 above)
Comment (izchen) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java]: Thanks for your review, I have modified the code

File: api/src/main/java/org/apache/iceberg/expressions/Literals.java (hunk 5 above)
Comment (izchen) [reply api/src/main/java/org/apache/iceberg/expressions/Literals.java]: Thanks for your review, I have modified the code

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java (hunk 6 above)
Comment (izchen) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java:661]: Thanks for your review. After testing, I think this is correct

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java (hunk 7 above)
Comment (izchen) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java]: Thanks for your review, I have modified the code

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/sql/TestSelect.java (hunk 8 above)
Comment (izchen) [reply spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/sql/TestSelect.java]: Thanks for your review, I have modified the code

File: api/src/main/java/org/apache/iceberg/util/ByteBuffers.java (hunk 9 above)
Comment (izchen) [reply api/src/main/java/org/apache/iceberg/util/ByteBuffers.java]: Thanks for your review, I have modified the code

**Comment by izchen** [comment]
> Eventually, we'll need to update more than just Spark v3.2. Additionally, it would be nice if Flink and even MR could be handled in this PR. 

### Flink

Executing this SQL test case in flink will report an error: 

```
[ERROR] Could not execute SQL statement. Reason:
org.apache.flink.table.api.ValidationException: Data type 'BINARY(2) NOT NULL' with conversion class '[B' does not support a value literal of class 'org.apache.calcite.avatica.util.ByteString'.
```

It seems that this is an internal error of flink, which has nothing to do with iceberg. I can add the following code to the flink code to solve this problem (I will try to submit a PR to the flink community in the future):

```scala
case BINARY =>
	// convert to byte[]
	literal.getValueAs(classOf[Array[Byte]])
```

https://github.com/apache/flink/blob/44378fa5fde6c17e1712a62b834cb6251605f416/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/RexNodeExtractor.scala#L400-L466

After using this code to fix flink, the SQL test case runs normally, returns the correct result, and the SQL description string in the flink UI is the correct X'110F'. I think this problem does not exist in Flink-runtime.

### Hive

According to the hive documentation, there is no binary type literal in hive. I think this problem does not exist in Hive-runtime.

### Trino

For Trino, I have not actually used Trino. Judging from the Trino code, Trino may not have this MR problem, but I do not have a Trino test environment. And we cannot modify the Trino-runtime code in this MR.

**Comment by izchen** [comment]
Sorry for the late reply. I left some feedback. Thank you for your review @kbendick @rdblue ！

**Comment by izchen** [comment]
@rdblue @kbendick Could you help review continue? 😄

**Comment by izchen** [comment]
I found a related trino community MR in the email. The current behavior of spark iceberg is used in this MR UT.
https://github.com/trinodb/trino/pull/10214

**Comment by izchen** [comment]
Do you think this PR is necessary? if not, I will close this PR.  Ping @rdblue @kbendick :)

File: spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java [hunk 10]
```diff
@@ -118,6 +119,8 @@ protected long waitUntilAfter(long timestampMillis) {
             return row.getList(pos);
           } else if (value instanceof scala.collection.Map) {
             return row.getJavaMap(pos);
+          } else if (value.getClass().isArray() && value.getClass().getComponentType().isPrimitive()) {
+            return IntStream.range(0, Array.getLength(value)).mapToObj(i -> Array.get(value, i)).toArray();
```
Comment (izchen) [inline spark/v3.2/spark/src/test/java/org/apache/iceberg/spark/SparkTestBase.java:123]: I think this is correct. The `toJava` method assumes that the elements of the returned array are all wrapper types, which is a false assumption. Spark's java api result has the potential to return an array of primitive types.

**Comment by rdblue** [comment]
This is fixed by #3728, which I have been reviewing lately and just committed. I didn't realize this was a separate PR until yesterday, so sorry that I hadn't come back to this one yet (I actually thought I was reviewing the same one!). In any case, the fix is nearly identical so this should be fixed by that other issue. Thanks, @izchen!
